be/src/vec/aggregate_functions/aggregate_function_java_udaf.h

```diff
@@ -0,0 +1,358 @@
+        return Status::OK();
+    }
+
+private:
+    jclass executor_cl;
+    jobject executor_obj;
+    jmethodID executor_ctor_id;
+
+    jmethodID executor_add_id;
+    jmethodID executor_merge_id;
+    jmethodID executor_serialize_id;
+    jmethodID executor_result_id;
+    jmethodID executor_close_id;
+
+    std::unique_ptr<int64_t[]> input_values_buffer_ptr;
+    std::unique_ptr<int64_t[]> input_nulls_buffer_ptr;
+    std::unique_ptr<int64_t[]> input_offsets_ptrs;
+    std::unique_ptr<int64_t> output_value_buffer;
+    std::unique_ptr<int64_t> output_null_value;
+    std::unique_ptr<int64_t> output_offsets_ptr;
+    std::unique_ptr<int64_t> output_intermediate_state_ptr;
+
+    bool first_init;
+    int argument_size = 0;
+    std::string serialize_data;
+};
+
+class AggregateJavaUdaf final
+        : public IAggregateFunctionDataHelper<AggregateJavaUdafData, AggregateJavaUdaf> {
+public:
+    AggregateJavaUdaf(const TFunction& fn, const DataTypes& argument_types, const Array& parameters,
+                      const DataTypePtr& return_type)
+            : IAggregateFunctionDataHelper(argument_types, parameters),
+              _fn(fn),
+              _return_type(return_type) {}
+    ~AggregateJavaUdaf() = default;
+
+    static AggregateFunctionPtr create(const TFunction& fn, const DataTypes& argument_types,
+                                       const Array& parameters, const DataTypePtr& return_type) {
+        return std::make_shared<AggregateJavaUdaf>(fn, argument_types, parameters, return_type);
+    }
+
+    void create(AggregateDataPtr __restrict place) const override {
+        new (place) Data(argument_types.size());
+        Status status = Status::OK();
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(_fn));
+    }
+
+    String get_name() const override { return _fn.name.function_name; }
+
+    DataTypePtr get_return_type() const override { return _return_type; }
+
+    // TODO: here calling add operator maybe only hava done one row, this performance may be poorly
+    // so it's possible to maintain a hashtable in FE, the key is place address, value is the object
+    // then we can calling add_bacth function and calculate the whole batch at once,
+    // and avoid calling jni multiple times.
+    void add(AggregateDataPtr __restrict place, const IColumn** columns, size_t row_num,
+             Arena*) const override {
+        this->data(place).add(columns, row_num, row_num + 1, argument_types);
+    }
+
+    // TODO: Here we calling method by jni, And if we get a thrown from FE,
+    // But can't let user known the error, only return directly and output error to log file.
+    void add_batch_single_place(size_t batch_size, AggregateDataPtr place, const IColumn** columns,
+                                Arena* arena) const override {
+        Status status = Status::OK();
+        RETURN_IF_STATUS_ERROR(status, data(place).init_udaf(_fn));
```

Comment:
use `RETURN_IF_ERROR`

Reply:
Is it necessary to `init_udaf` for each time to call `add_batch_single_place`? Seems you have called it in `create`.